### PR TITLE
Add support to redirect from a Hive catalog to an Iceberg catalog

### PR DESCRIPTION
**Pull request by lxynov**
**Add catalog redirection support to SPI**
Add a catalog redirection method to connector SPI. It allows a connector
to redirect to another catalog which may or may not use the same
connector.

Catalog redirection is tried whenever a table or view is accessed, e.g.,
in SELECT, INSERT, DROP TABLE, ALTER TABLE, GRANT, etc., queries. However,
it's not tried with procedures.

Multi-stop redirection without loop is allowed within a max depth.

**Add support to redirect from Hive to an Iceberg catalog**
When enabled, Hive Connector would redirect a table access to an Iceberg
catalog when the table is an Iceberg table.

----
Issue: #4442 
Umbrella issue: #1324 

Some notes:
 - Access control is enforced on the eventual table.
 - Information schema queries (`show tables`, `select * from information_schema.columns`) stay the same. They won't include results from other catalogs. Having said that, in the Hive/Iceberg case, a Hive catalog does include results of Iceberg tables inherently.
 - I thought about adding warning messages for users. Can work on that as a follow-up if it makes sense.

### CODE REVIEW

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 1]
```diff
@@ -257,6 +261,8 @@
 import static java.util.stream.Collectors.toSet;
 import static org.apache.hadoop.hive.metastore.TableType.EXTERNAL_TABLE;
 import static org.apache.hadoop.hive.metastore.TableType.MANAGED_TABLE;
+import static org.apache.iceberg.BaseMetastoreTableOperations.ICEBERG_TABLE_TYPE_VALUE;
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:264]: Since these are the only usages of the Iceberg library, it would be better to manually add the constants rather than a dependency on Iceberg (avoids any potential conflicts or other issues).

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java [hunk 2]
```diff
@@ -301,6 +307,12 @@
     private final HiveStatisticsProvider hiveStatisticsProvider;
     private final AccessControlMetadata accessControlMetadata;
 
+    // Copied from IcebergTableHandle
+    private static final Pattern ICEBERG_TABLE_NAME_PATTERN = Pattern.compile("" +
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:311]: Should we simplify this so that we don't depend on the exact Iceberg table name format? Something like `([^$@]+)([$@].+)?` Or maybe it's better to keep it exactly the same.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSessionProperties.java [hunk 3]
```diff
@@ -91,6 +91,8 @@
     private static final String QUERY_PARTITION_FILTER_REQUIRED = "query_partition_filter_required";
     private static final String PROJECTION_PUSHDOWN_ENABLED = "projection_pushdown_enabled";
     private static final String PARQUET_OPTIMIZED_WRITER_ENABLED = "parquet_optimized_writer_enabled";
+    private static final String REDIRECTING_TO_ICEBERG_ENABLED = "redirecting_to_iceberg_enabled";
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSessionProperties.java:94]: We could use a single property `redirect_to_iceberg` that is the catalog name. If it's not set, we don't redirect. Thoughts?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 1 above)
Comment (lxynov) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:264]: Yeah, agreed. I thought about this, too. They are unlikely to change as they are in the spec.

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java (hunk 2 above)
Comment (lxynov) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java:311]: Hmm I'd vote for keeping it exactly the same

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSessionProperties.java (hunk 3 above)
Comment (lxynov) [reply presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSessionProperties.java:94]: I think having `REDIRECTING_TO_ICEBERG_ENABLED` can make things clearer to users. They can set it false explicitly if they want to fall back to use the Hive table.

File: presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java [hunk 4]
```diff
@@ -1796,6 +1796,28 @@ public AnalyzePropertyManager getAnalyzePropertyManager()
...
+    @Override
+    public Optional<String> redirectCatalog(Session session, QualifiedObjectName tableName)
+    {
+        requireNonNull(tableName, "tableName is null");
+
+        if (tableName.getCatalogName().isEmpty() || tableName.getSchemaName().isEmpty() || tableName.getObjectName().isEmpty()) {
+            return Optional.empty();
+        }
+
+        Optional<CatalogMetadata> catalog = getOptionalCatalogMetadata(session, tableName.getCatalogName());
```
[the remark this answers is not in the document]
Comment (lxynov) [reply presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java:1808]: `metadata.redirectCatalog(connectorSession, tableName.asSchemaTableName())` returns `Optional<String>` instead of `String`, making it difficult to follow the pattern here.

**Comment by lxynov** [comment]
@electrum Thanks for the review. I've updated the PR

**Comment by lxynov** [comment]
> This shouldn't be Iceberg specific. One could want to redirect to some other catalog for some other reasons.
eg HBase table to Phoenix, Hive JDBC table to respective JDBC connector, etc.

@findepi  People can implement the `ConnectorMetadata:: redirectCatalog` method to achieve these

**Comment by findepi** [comment]
@lxynov great! I actually mean redirecting from hive to other connectors.
This needs to be extensible. Therefore, we should not have iceberg specific 
configuration properties like `hive.redirecting-to-iceberg-enabled`, as they
clearly prevent extensibility.

**Comment by lxynov** [comment]
@findepi oh I see your point. That makes sense. We can name it `hive.catalog-redirection-enabled`. @electrum what do you think?

**Comment by electrum** [comment]
I'm not sure it makes sense to be extensible at the user level. The redirection here is specific to Iceberg based on the Hive connector's knowledge that the table is an Iceberg table. If we had an HBase connector in addition to a Phoenix connector, it seems that the same level of specific knowledge would apply.

File: presto-hive/pom.xml [hunk 5]
```diff
@@ -80,7 +80,7 @@
         <dependency>
             <groupId>io.airlift</groupId>
             <artifactId>jmx</artifactId>
-        </dependency>
```
Comment (electrum) [inline presto-hive/pom.xml:83]: This change is unrelated

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveConfig.java [hunk 6]
```diff
@@ -134,6 +134,9 @@
 
     private boolean projectionPushdownEnabled = true;
 
+    private boolean redirectingToIcebergEnabled;
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveConfig.java:137]: This would sound better written as "redirect" so the phrase is "redirect to Iceberg". I failed to come up with a precise reason why. Perhaps because "redirecting" is a gerund (verb functioning as a noun) rather than a noun.

This applies to all usages (config and session property).

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveConfig.java [hunk 7]
```diff
@@ -977,4 +980,31 @@ public HiveConfig setProjectionPushdownEnabled(boolean projectionPushdownEnabled
         this.projectionPushdownEnabled = projectionPushdownEnabled;
         return this;
     }
+
+    public boolean isRedirectingToIcebergEnabled()
+    {
+        return redirectingToIcebergEnabled;
+    }
+
+    @Config("hive.redirecting-to-iceberg-enabled")
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveConfig.java:989]: Let's name these
* `hive.redirect-to-iceberg.enabled`
* `hive.redirect-to-iceberg.catalog`

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSessionProperties.java [hunk 8]
```diff
@@ -369,6 +371,16 @@ public HiveSessionProperties(
...
                         parquetWriterConfig.isParquetOptimizedWriterEnabled(),
+                        false),
+                booleanProperty(
+                        REDIRECTING_TO_ICEBERG_ENABLED,
+                        "Enable redirecting to a catalog configured with Iceberg Connector",
+                        hiveConfig.isRedirectingToIcebergEnabled(),
+                        false),
+                stringProperty(
+                        REDIRECTING_ICEBERG_CATALOG,
+                        "The Iceberg catalog to redirect to",
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSessionProperties.java:382]: > The target Iceberg catalog for redirection

File: presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSessionProperties.java [hunk 9]
```diff
@@ -91,6 +91,8 @@
     private static final String QUERY_PARTITION_FILTER_REQUIRED = "query_partition_filter_required";
     private static final String PROJECTION_PUSHDOWN_ENABLED = "projection_pushdown_enabled";
     private static final String PARQUET_OPTIMIZED_WRITER_ENABLED = "parquet_optimized_writer_enabled";
+    private static final String REDIRECTING_TO_ICEBERG_ENABLED = "redirecting_to_iceberg_enabled";
+    private static final String REDIRECTING_ICEBERG_CATALOG = "redirecting_iceberg_catalog";
```
Comment (electrum) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/HiveSessionProperties.java:95]: `redirect_to_iceberg_catalog`

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveRedirectionToIceberg.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+import static io.prestosql.tempto.query.QueryExecutor.query;
+import static io.prestosql.tests.TestGroups.HIVE_REDIRECTION_TO_ICEBERG;
+import static io.prestosql.tests.TestGroups.PROFILE_SPECIFIC_TESTS;
+import static io.prestosql.tests.utils.QueryExecutors.onPresto;
+import static java.lang.String.format;
+
+public class TestHiveRedirectionToIceberg
+        extends HiveProductTest
+{
+    private static final String CREATE_ICEBERG_TABLE_TEAMPLATE = "CREATE TABLE %s (_string VARCHAR, _bigint BIGINT, _date DATE) WITH (partitioning = ARRAY['_bigint', '_date'])";
```
Comment (electrum) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveRedirectionToIceberg.java:37]: Typo "teamplate" -> template

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveRedirectionToIceberg.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+    private static final String CREATE_HIVE_TABLE_TEAMPLATE = "CREATE TABLE %s (_string VARCHAR, _bigint BIGINT, _date DATE) WITH (partitioned_by = ARRAY['_bigint', '_date'])";
+    private static final String INSERT_TEMPLATE = "INSERT INTO %s VALUES " +
+            "(NULL, NULL, NULL), " +
+            "('abc', 1, DATE('2020-08-04')), " +
```
Comment (electrum) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveRedirectionToIceberg.java:41]: No need for parenthesis
```sql
DATE '2020-08-04'

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveRedirectionToIceberg.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        onPresto().executeQuery(format(CREATE_ICEBERG_TABLE_TEAMPLATE, "iceberg.default.test_select_drop"));
+        onPresto().executeQuery(format(INSERT_TEMPLATE, "iceberg.default.test_select_drop"));
+        assertThat(onPresto().executeQuery("SELECT * FROM iceberg.default.test_select_drop")).containsOnly(EXPECTED_ROWS);
+        assertThat(onPresto().executeQuery("SELECT * FROM hive.default.test_select_drop")).containsOnly(EXPECTED_ROWS);
+        assertThat(onPresto().executeQuery("SELECT _bigint, _date FROM hive.default.\"test_select_drop$partitions\"")).containsOnly(
+                row(null, null),
+                row(1, Date.valueOf("2020-08-04")),
+                row(2, Date.valueOf("2020-08-04")));
+        onPresto().executeQuery("DROP TABLE hive.default.test_select_drop");
+        assertThat(() -> query("SELECT * FROM iceberg.default.test_select_drop"))
```
Comment (electrum) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveRedirectionToIceberg.java:61]: Maybe also verify that select from Hive fails

File: presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveRedirectionToIceberg.java [hunk 13]
```diff
@@ -0,0 +1,129 @@
...
+        assertThat(onPresto().executeQuery("SELECT * FROM iceberg.default.test_insert")).containsOnly(EXPECTED_ROWS);
+        assertThat(onPresto().executeQuery("SELECT * FROM hive.default.test_insert")).containsOnly(EXPECTED_ROWS);
+        onPresto().executeQuery("DROP TABLE iceberg.default.test_insert");
+    }
+
+    @Test(groups = {HIVE_REDIRECTION_TO_ICEBERG, PROFILE_SPECIFIC_TESTS})
+    public void testDescribe()
+    {
+        onPresto().executeQuery(format(CREATE_ICEBERG_TABLE_TEAMPLATE, "iceberg.default.test_describe"));
+        assertThat(onPresto().executeQuery("DESCRIBE hive.default.test_describe"))
```
Comment (electrum) [inline presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveRedirectionToIceberg.java:79]: Perhaps unrelated to this PR, but what happens if we create with an Iceberg type like `TIME` that Hive doesn't support? (`UUID` would be another example but @rdblue says that type should be removed)

Comment (lxynov) [reply presto-product-tests/src/main/java/io/prestosql/tests/hive/TestHiveRedirectionToIceberg.java:79]: I believe we would be able to create such a table and query it through Iceberg Connector, but unable to query it through Hive Connector. (https://github.com/prestosql/presto/blob/6e43e24194f7d3cb61d5222dd7c9ccafdc4f4873/presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/HiveTableOperations.java and https://github.com/apache/iceberg/blob/c36740be1eed8f1773b9f48bcf498f56dc784d84/hive/src/main/java/org/apache/iceberg/hive/HiveTypeConverter.java)
Since this PR's implementation redirects the query to Iceberg Connector entirely, people can query it as `hive.default.test_describe`, too

**Comment by lxynov** [comment]
@electrum Thanks for the review! PR updated. The failing test seems unrelated.

**Review by dain — COMMENTED** [review]
Is there going to be a redirect Iceberg to Hive feature also?

File: presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java [hunk 14]
```diff
@@ -965,4 +965,12 @@ default ConnectorTableProperties getTableProperties(ConnectorSession session, Co
      * <p>
      */
     default void validateScan(ConnectorSession session, ConnectorTableHandle handle) {}
+
+    /**
+     * Redirects to another catalog which may or may not use the same connector
+     */
+    default Optional<String> redirectCatalog(ConnectorSession session, SchemaTableName tableName)
```
Comment (dain) [inline presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:972]: I expect this to return a new FQN.  I know is the cases we are dealing with are all simple "change the catalog" case, but a change to `CatalogSchemTableName` would be much more powerful here and support a semantic catalog.  It is also easy to convert from this current code to a FQN version.

Comment (electrum) [reply presto-spi/src/main/java/io/prestosql/spi/connector/ConnectorMetadata.java:972]: We intentionally do not rewrite the schema or table name here, as that makes operations like renames nonsensical. What you are describing is a completely different feature that we considered but rejected in favor of the catalog redirection approach. I agree that we want both, eventually, but the semantics of each need to be carefully considered. The restriction to only allow catalog here actually makes it more powerful in that it can support features like rename.

**Comment by electrum** [comment]
Replaced by #5160
